op-supervisor/supervisor/backend/backend.go

```diff
@@ -60,61 +68,115 @@ func NewSupervisorBackend(ctx context.Context, logger log.Logger, m Metrics, cfg
 	if err != nil {
 		return nil, fmt.Errorf("failed to load dependency set: %w", err)
 	}
+	chains := depSet.Chains()
 
-	// create the chains db
-	chainsDB := db.NewChainsDB(logger)
-
-	// create an empty map of chain monitors
-	chainProcessors := make(map[types.ChainID]*processors.ChainProcessor, len(cfg.L2RPCs))
+	// create initial per-chain resources
+	chainsDBs := db.NewChainsDB(logger)
+	chainProcessors := make(map[types.ChainID]*processors.ChainProcessor, len(chains))
+	chainMetrics := make(map[types.ChainID]*chainMetrics, len(chains))
 
 	// create the supervisor backend
 	super := &SupervisorBackend{
 		logger:          logger,
 		m:               m,
 		dataDir:         cfg.Datadir,
 		depSet:          depSet,
+		chainDBs:        chainsDBs,
 		chainProcessors: chainProcessors,
-		db:              chainsDB,
+		chainMetrics:    chainMetrics,
+		// For testing we can avoid running the processors.
+		synchronousProcessors: cfg.SynchronousProcessors,
+	}
+
+	// Initialize the resources of the supervisor backend.
+	// Stop the supervisor if any of the resources fails to be initialized.
+	if err := super.initResources(ctx, cfg); err != nil {
+		err = fmt.Errorf("failed to init resources: %w", err)
+		return nil, errors.Join(err, super.Stop(ctx))
 	}
 
-	// from the RPC strings, have the supervisor backend create a chain monitor
-	// don't start the monitor yet, as we will start all monitors at once when Start is called
+	return super, nil
+}
+
+// initResources initializes all the resources, such as DBs and processors for chains.
+// An error may returned, without closing the thus-far initialized resources.
+// Upon error the caller should call Stop() on the supervisor backend to clean up and release resources.
+func (su *SupervisorBackend) initResources(ctx context.Context, cfg *config.Config) error {
+	chains := su.depSet.Chains()
+
+	// for each chain known to the dependency set, create the necessary DB resources
+	for _, chainID := range chains {
+		if err := su.openChainDBs(chainID); err != nil {
+			return fmt.Errorf("failed to open chain %s: %w", chainID, err)
+		}
+	}
+
+	// for each chain initialize a chain processor service
+	for _, chainID := range chains {
+		logProcessor := processors.NewLogProcessor(chainID, su.chainDBs)
+		chainProcessor := processors.NewChainProcessor(su.logger, chainID, logProcessor, su.chainDBs)
+		su.chainProcessors[chainID] = chainProcessor
+	}
+
+	// the config has some RPC connections to attach to the chain-processors
 	for _, rpc := range cfg.L2RPCs {
-		err := super.addFromRPC(ctx, logger, rpc, false)
+		err := su.attachRPC(ctx, rpc)
 		if err != nil {
-			return nil, fmt.Errorf("failed to add chain monitor for rpc %v: %w", rpc, err)
+			return fmt.Errorf("failed to add chain monitor for rpc %v: %w", rpc, err)
 		}
 	}
```

Comment:
Every chain in the Supervisor is maintained by the backend in ~2 ways:
* Database Storage
* RPC Connection

The databases can be generated per chain as soon as we know about the chain, as the chainID is the only important part. However, the RPC connections are not always ready. We found this was an issue in testing, a circular bootstrap requirement where the Nodes and Supervisor must be aware of one-another.

So, now we make databases for *all* chains, and we connect to any RPCs specified in the config. As we connect to them, we discover which chain they are for and connect them. And you can use the AddRPC API to add more connections later.

These RPC connections are how the Supervisor gets logs. The Nodes push updates ("I have a new Unsafe block"), and the Supervisor will respond by pulling the receipts for that block using the RPC for that chain.

@protolambda: note, there's an inconsistency risk here, where the node informing you of an update might *not* be the node you connect to via RPC. you could discover that the RPC node hasn't gotten the block yet. For now we can solve this by just saying "don't do that, use one node only".